compiler/frontend/src/org/jetbrains/kotlin/resolve/calls/smartcasts/DataFlowUtils.kt

```diff
@@ -40,13 +40,13 @@ fun KtExpression?.getKotlinTypeWithPossibleSmartCastToFP(
     if (descriptor != null) {
         val dataFlow = dataFlowValueFactory.createDataFlowValue(this, givenType, bindingContext, descriptor)
         val stableTypes = bindingContext.getDataFlowInfoBefore(this).getStableTypes(dataFlow, languageVersionSettings)
-        stableTypes.firstNotNullResult {
+        return stableTypes.firstNotNullResult {
             when {
                 KotlinBuiltIns.isDoubleOrNullableDouble(it) -> it
                 KotlinBuiltIns.isFloatOrNullableFloat(it) -> it
                 else -> null
             }
-        }?.let { return it }
+        } ?: stableTypes.firstOrNull() ?: givenType
```

Comment:
I'm afraid we can break something in back-end with this change, because this function has the second use-site in back-end. Also, this change contradicts with the function name. We should keep back-end behavior here. I'd consider adding a lambda as the last function parameter, which takes `givenType` and `stableTypes` as arguments and returns something like `defaultType`. By default this lambda should return `givenType`, but in inspection we can change its value to `stableTypes.firstOrNull() ?: givenType`.